### PR TITLE
release/v0.6.0

### PR DESCRIPTION
**Pull request by cmccormick1**
Includes implementing MSC's error_prop functions and plotting temperature and metallicity error bars for composite measurements (issue #20).

### CODE REVIEW

File: general.py [hunk 1]
```diff
@@ -134,13 +131,14 @@ def run_bin_analysis(err_prop = False, indiv = False):
     
     #Run validation table
     valid_table.make_validation_table(fitspath)
+    valid_file = fitspath + filename_dict['bin_valid']
```
Comment (astrochun) [inline general.py:134]: You might consider using `os.path.join` to define paths.  This handles the extraneous `/` in path.
`from os.path import join`
`valid_file = join(fitspath, filename_dict['bin_valid'])`

File: plots.py [hunk 2]
```diff
@@ -51,36 +51,27 @@ def bin_derived_props_plots(fitspath, metal_file, em_file, bin_file, valid_file,
...
-        metal_low_err = []
-        metal_high_err = []
-        for ii in range(len(Te_err)):
-            Te_low_err.append(Te_err[ii][0])
-            Te_high_err.append(Te_err[ii][1])
-            metal_low_err.append(metal_err[ii][0])
-            metal_high_err.append(metal_err[ii][1])
+        Te_low_err = [Te_err[ii][0] for ii in range(len(Te_err))]
+        Te_high_err = [Te_err[ii][1] for ii in range(len(Te_err))]
+        metal_low_err = [metal_err[ii][0] for ii in range(len(metal_err))]
```
Comment (astrochun) [inline plots.py:65]: I think you can do: `metal_low_err = metal_err[:,0]` and similarly for `metal_high_err`

File: plots.py [hunk 3]
```diff
@@ -283,6 +273,23 @@ def indiv_derived_props_plots(fitspath, dataset, restrict_MTO = False, revised =
...
         LHb_metal_tbl = asc.read(fitspath + 'mass_LHbeta_bin/' + dataset + filename_dict['bin_derived_prop'])
+        
+    if err_bars == True:
+        der_prop_err_M = np.load(fitspath + 'massbin/' + dataset + npz_filename_dict['der_prop_errors'])
+        der_prop_err_LHb = np.load(fitspath + 'massbin/' + dataset + npz_filename_dict['der_prop_errors'])
+        
+        metal_err_M = der_prop_err_M['12+log(O/H)_lowhigh_error']
+        metal_err_LHb = der_prop_err_LHb['12+log(O/H)_lowhigh_error']
+        
+        metal_low_err_M = [metal_err_M[ii][0] for ii in range(len(metal_err_M))]
```
Comment (astrochun) [inline plots.py:284]: It seems that you do this often, extract one of the errorbars.  You might consider creating a function that provides both lower and upper errors.

File: plots.py [hunk 4]
```diff
@@ -567,11 +574,12 @@ def indiv_derived_props_plots(fitspath, dataset, restrict_MTO = False, revised =
...
     ax11[0].scatter(mass_avg_mass[mass_detect], mass_metal[mass_detect], s = 25, color = 'red', label = 'Bin Detections')
     ax11[0].scatter(mass_avg_mass[mass_nondetect], mass_metal[mass_nondetect], s = 25, color = 'red', marker = '^', label = 'Bin Non-Detections', alpha = 0.5)
-    #ax11[0].errorbar(mass_avg_mass[mass_detect], mass_metal[mass_detect], yerr = , s = 25, color = 'red', label = 'Bin Detections')
-    #ax11[0].errorbar(mass_avg_mass[mass_nondetect], mass_metal[mass_nondetect], yerr = , s = 25, color = 'red', marker = '^', label = 'Bin Non-Detections', alpha = 0.5)
     #HBeta bin detections and non-detections
     ax11[1].scatter(LHb_avg_mass[LHb_detect], LHb_metal[LHb_detect], s = 25, color = 'red', label = 'Bin Detections')
     ax11[1].scatter(LHb_avg_mass[LHb_nondetect], LHb_metal[LHb_nondetect], s = 25, color = 'red', marker = '^', label = 'Bin Non-Detections', alpha = 0.5)
+    if err_bars == True:
+        ax11[0].errorbar(mass_avg_mass[mass_detect], mass_metal[mass_detect], yerr = metal_lowhigh_err_M, fmt = '.')
+        ax11[0].errorbar(LHb_avg_mass[LHb_detect], LHb_metal[LHb_detect], yerr = metal_lowhigh_err_LHb, fmt = '.')
```
Comment (astrochun) [inline plots.py:582]: I think there's a mistake in line 582 as it should be `ax11[1]`

File: general.py [hunk 5]
```diff
@@ -163,16 +161,11 @@ def run_bin_analysis(err_prop = False, indiv = False):
...
-                        fitspath + 'metal_xpeaks.npz', fitspath + 'metallicity_pdf.npz', fitspath + 'Te_errors.npz']
-        FR_dict_list = [fitspath + 'flux_propdist.npz', fitspath + 'flux_errors.npz', fitspath + 'flux_xpeak.npz']
-        histogram_plots.run_histogram_TM(fitspath, metal_file, TM_dict_list, valid_file, sharex=False)
-        histogram_plots.run_histogram_FR(fitspath, em_file, FR_dict_list, valid_file, sharex=False)
+        error_prop.fluxes_derived_prop(fitspath, binned_data = True)
         metal_file = fitspath + filename_dict['bin_derived_prop_rev']
         em_file = fitspath + filename_dict['bin_fit_rev']
         out_fname = fitspath + filename_dict['bin_derived_prop_rev'].replace('.tbl', '.pdf')
-        plots.bin_derived_props_plots(fitspath, metal_file, em_file, bin_file, valid_file, out_fname, hbeta_bin = bool_hbeta_bin)
+        plots.bin_derived_props_plots(fitspath, metal_file, em_file, bin_file, valid_file, out_fname, hbeta_bin = bool_hbeta_bin, err_bars = True)
```
Comment (astrochun) [inline general.py:168]: Should there handling for `error_prop == False`?

File: plots.py [hunk 6]
```diff
@@ -51,36 +51,27 @@ def bin_derived_props_plots(fitspath, metal_file, em_file, bin_file, valid_file,
...
-            Te_low_err.append(Te_err[ii][0])
-            Te_high_err.append(Te_err[ii][1])
-            metal_low_err.append(metal_err[ii][0])
-            metal_high_err.append(metal_err[ii][1])
+        Te_low_err = [Te_err[ii][0] for ii in range(len(Te_err))]
+        Te_high_err = [Te_err[ii][1] for ii in range(len(Te_err))]
+        metal_low_err = [metal_err[ii][0] for ii in range(len(metal_err))]
+        metal_high_err = [metal_err[ii][1] for ii in range(len(metal_err))]
+        
+        Te_low_err = np.log10(1 + (Te_low_err)/T_e[detect])
```
Comment (astrochun) [inline plots.py:68]: I believe for low error this should be
`Te_low_err = -1*np.log10(1 - Te_low_err/T_e[detect])`

File: general.py [hunk 7]
```diff
@@ -163,16 +161,11 @@ def run_bin_analysis(err_prop = False, indiv = False):
     
     #Run error propagation, histograms, and revised data plots
     if err_prop == True:
```
Comment (astrochun) [inline general.py:163]: Note that you can simplify to: `if err_prop:` (this handles boolean true/false or binary 1/0)

File: plots.py [hunk 8]
```diff
@@ -283,6 +273,23 @@ def indiv_derived_props_plots(fitspath, dataset, restrict_MTO = False, revised =
     else:    
         mass_metal_tbl = asc.read(fitspath + 'massbin/' + dataset + filename_dict['bin_derived_prop'])
         LHb_metal_tbl = asc.read(fitspath + 'mass_LHbeta_bin/' + dataset + filename_dict['bin_derived_prop'])
+        
+    if err_bars == True:
+        der_prop_err_M = np.load(fitspath + 'massbin/' + dataset + npz_filename_dict['der_prop_errors'])
+        der_prop_err_LHb = np.load(fitspath + 'massbin/' + dataset + npz_filename_dict['der_prop_errors'])
```
Comment (astrochun) [inline plots.py:279]: `os.path.join` would make this easier as it accepts multiple folders as separate inputs:
`join(fitspath, 'massbin', dataset, npz_filename_dict['der_prop_errors'])`

File: plots.py (hunk 3 above)
Comment (astrochun) [reply plots.py:284]: There might also be a way of using `numpy.transpose` to flip the error arrays for `matplotlib`

**Review by astrochun — CHANGES_REQUESTED** [review]
See comments for suggestions and possible typo

File: plots.py (hunk 3 above)
Comment (cmccormick1) [reply plots.py:284]: Yes, I was considering making a separate function for extracting error bars, but I was going to change it on my feature/plots_reorganize branch.

**Pull request merged**